src/Import/Remote/Composer.php

```diff
@@ -58,38 +61,48 @@ public function update(bool $force = false, bool $displayProgress = true): void
     {
         $dir = PathHelper::getRoot() . self::VENDOR_DIR;
 
-        if ($force || !$this->isInstalled($dir)) {
-            $this->filesystem->mkdir($dir);
+        if (!($force || !$this->isInstalled($dir))) {
+            $this->logger->debug('Packages were already required, no need to run Composer.');
 
-            file_put_contents($dir . '.gitignore', "*\n");
-            file_put_contents("{$dir}/composer.json", json_encode($this->configuration, \JSON_PRETTY_PRINT | \JSON_THROW_ON_ERROR));
+            return;
+        }
 
-            $progressIndicator = null;
-            if ($displayProgress) {
-                $progressIndicator = new ProgressIndicator($this->output, null, 100, ['⠏', '⠛', '⠹', '⢸', '⣰', '⣤', '⣆', '⡇']);
-                $progressIndicator->start('<comment>Downloading remote packages</comment>');
-            }
+        $this->filesystem->mkdir($dir);
 
-            $this->run(['update'], callback: function () use ($progressIndicator) {
-                if ($progressIndicator) {
-                    $progressIndicator->advance();
-                }
-            });
+        file_put_contents($dir . '.gitignore', "*\n");
+        file_put_contents("{$dir}/composer.json", json_encode($this->configuration, \JSON_PRETTY_PRINT | \JSON_THROW_ON_ERROR));
 
+        $progressIndicator = null;
+        if ($displayProgress) {
+            $progressIndicator = new ProgressIndicator($this->output, null, 100, ['⠏', '⠛', '⠹', '⢸', '⣰', '⣤', '⣆', '⡇']);
+            $progressIndicator->start('<comment>Downloading remote packages</comment>');
+        }
+
+        $this->run(['update'], callback: function () use ($progressIndicator) {
             if ($progressIndicator) {
-                $progressIndicator->finish('<info>Remote packages imported</info>');
+                $progressIndicator->advance();
             }
-            $this->writeInstalled($dir);
-        } else {
-            $this->logger->debug('Packages were already required, no need to run Composer.');
+        });
+
+        if ($progressIndicator) {
+            $progressIndicator->finish('<info>Remote packages imported</info>');
         }
+
+        $this->writeInstalled($dir);
+
+        $this->hasRun = true;
```

Comment:
can we directly throw here instead of tracking this boolean ?

Reply:
yes and no. No because we can have multiple (sequentially) composer up. If you require A, and A require B. So let's restart only once to avoid increasing memory footprint